README.md
# Bomb'n'Break for SNES

40 years since the NES came out. Celebrating this anniversary and as I found a nice framework to code for
the SNES in C I tried to write a little game myself.

<img src="doc/screenshot.png" alt="game screenshot">

# Building

Dependencies:
- [pvsneslib](https://github.com/alekmaul/pvsneslib)
- Unix shell
- Python 3.x
- make

Make sure to have the environment variable `PVSNESLIB_HOME` point to the installation directory of pvsneslib.

Default debug build:
```sh
make
```

## Build Options

|Option     |Description                                             |
|-----------|--------------------------------------------------------|
|NDEBUG=1   |Disable debug information (i.e. release build).         |
|USE_NTSC=1 |Assume NTSC instead of PAL (default).                   |
|HAS_BGM=1  |Enable background music (requires `res/bgm1.it`)        |
|HAS_SFX=1  |Enable explosion sound effect (requires `res/sfx1.wav`) |

# BGM/SFX

For private means I am using the explosion effect from BlitzBlaster
by Sergio Marcello and the track Jazzy Nork by Crystal Symphonies
which are not included here due to missing license agreements.

You can include your own sounds performing the following steps.

## Adding explosion SFX

- Create a [WAV file](https://en.wikipedia.org/wiki/WAV) with PCM, 11025 Hz and 8-bits per sample.
- Store the file in `res/sfx1.wav`.
- Adjust the size passed to `spcAllocateSoundRegion()` in `main.c`.
- Run `make` with `HAS_SFX=1` as command-line parameter.

## Adding BGM

- Create an [IT file](https://de.wikipedia.org/wiki/Impulse_Tracker) with instruments
  instead of samples using 8-bits per sample. I recommend using [Schism Tracker](https://schismtracker.org/).
- Store the file in `res/bgm1.it`.
- Adjust the number of used soundbanks with `spcSetBank()` in `main.c` (using reverse order).
- Run `make` with `HAS_BGM=1` as command-line parameter.

# Development

The code is written in C and [65816 assembly](https://github.com/nArnoSNES/wla-dx-65816).  
Many considerations have been done to avoid performance issues. E.g. the linear random number
generator is written in assembly. This chapter summaries some development insights and design
considerations.

## Tools

Recommended tools for development:

- [pvsneslib](https://github.com/alekmaul/pvsneslib)
- [bsnes](https://github.com/bsnes-emu/bsnes) or [Mesen](https://www.mesen.ca)
- [M1TE](https://github.com/nesdoug/M1TE2) (Windows)
- [Schism Tracker](https://schismtracker.org/)
- [Gimp](https://www.gimp.org/)
- [Photoshop](https://www.adobe.com/products/photoshop.html)
- [MSYS2](https://www.msys2.org/) (Windows)

Note that `smconv` crashed trying to convert the IT file written by [OpenMPT](https://openmpt.org/) for me.

## References

- [65816 Reference](https://wiki.superfamicom.org/65816-reference)
- [SNES Assembly](https://ersanio.gitbook.io/assembly-for-the-snes/)
- [WLA-DX-65816](https://github.com/nArnoSNES/wla-dx-65816)
- [SNES Development Wiki](https://wiki.superfamicom.org/)
- [Hardware Registers](https://media.smwcentral.net/Ersanio/SMWCstuff/Advanced%20documentation/qsnesdoc.html)
- [DMA & HDMA Explained](https://www.youtube.com/watch?v=K7gWmdgXPgk)

## Background, Foreground, Sprites

The background plane is separated in a background (`BG_NR`) and foreground (`FG_NR`) layer.
The identifiers are chosen according to the [background priorities in mode 1](https://wiki.superfamicom.org/backgrounds).
Both layers consist of 64x32 tiles á 8x8 pixels. The window resolution is only 32x28 tiles
and 32x32 tiles are used per page in the SNES to simplifies various operations. Two pages
are used to make the slide-in/slide-out effect. The hidden page is being replaced before the
next one is being shown.  
The tiles form the title screen, options screen and the game screen completely. Only the
player characters are being drawn using sprites.  
The title screen and option screen are being modified directly in VRAM whereas the game screen
uses the WRAM buffers `gameFieldLow` for tile indices and and `gameFieldHigh` for attribute
data. These are being updated after each VBlank if `refreshGameScreenLow`, respective
`refreshGameScreenHigh` are set `true` using DMA transfer. This avoids graphic glitches.
The additional arrays `aniField` and `ttlField` are being used to perform the animation of
explosion and breaking walls correctly. The animation data for the bombs is being handled in
the `bombList` array associated to `p1` and `p2`.  
Note that `fieldElemIndex` and `fTypeMap` are being used to speed-up animation and collision checks.

The following table shows how the tile maps, tiles and palettes are being used (see also `src/data.asm`).

|File        |Identifiers               |Type      |Use in/as                                          |
|------------|--------------------------|----------|---------------------------------------------------|
|bg1.chr     |bg1Tiles, bg1TilesEnd     |tiles     |Title screen background tiles.                     |
|bg1.map     |bg1Map, bg1MapEnd         |tile map  |Title screen background map (title/start).         |
|bg1.pal     |bg1Pal, bg1PalEnd         |palette   |Title screen background palette number #1.         |
|bg2.map     |bg2Map, bg2MapEnd         |tile map  |Game screen background map.                        |
|fg1.chr     |fg1Tiles, fg1TilesEnd     |tiles     |Title screen foreground tiles.                     |
|fg1.map     |fg1Map, fg1MapEnd         |tile map  |Title screen foreground map (credits).             |
|fg1.pal     |fg1Pal, fg1PalEnd         |palette   |Title screen foreground palette number #2.         |
|fg2.chr     |fg2Tiles, fg2TilesEnd     |tiles     |Option/game screen foreground tiles.               |
|fg2.pal     |fg2Pal, fg2PalEnd         |palette   |Option/game screen foreground palette number #3.   |
|options.map |optionsMap, optionsMapEnd |tile map  |Option screen foreground map with the options.     |
|field.map   |fieldMap, fieldMapEnd     |tile map  |Initial game screen foreground map with the field. |
|p12.chr     |p12Tiles, p12TilesEnd     |tiles     |Player 1 and 2 sprite tiles.                       |
|p12.pal     |p12Pal, p12PalEnd         |palettes  |Player 1 (#4) and 2 (#5) sprite palettes.          |

`options.map` and `field.map` are using the tiles and palette of `fg2.chr` with priority flag set to 1.  
`fg2.chr` also contains the ASCII characters used to display the numeric values on the option and game screen.

## Time Handling

NTSC (60 Hz) or PAL (50 Hz) is being selected via build switch. Knowing their frame rate it is easy
to use VBlanks to count the time. `FP10HZ` is being defined accordingly to simplify this task. The macro
holds the number of frames per 1/10s. All animations are based on this frequency of 10 Hz.  
The time base is being used in `clickDelay()` to avoid fast value changes in the option screen and
for the animations and game time handling in `handleGame()`. The speed of the slide-in/slide-out animation
is determined from the value of `SLIDE_SPEED` which is also derived from the video mode.  
See also the global variables `framesUntil10Hz`, `counter10Hz` and `untilSecond` used for global time management.

## Collision Detection

Collision detection is simply done via bounding box using the 4 corners of the player sprite defined via
`P_LEFT`, `P_RIGHT`, `P_TOP` and `P_BOTTOM` with `P_MID_X` and `P_MID_Y` forming its center used for bomb
placement. It is bing checked whether one of these corners falls within a given game field or not.
To detect whether the most recently dropped bomb field has been left, it is being checked whether the
16x16 fields around these four corners contain the dropped bomb field or not.
No hitmask or other technique is being used. Most of the collision events are being tested using the game
field coordinates.

## Variable Storage

Most of the variables are being stored as global variables due to the fact that the used CPU provides
only a 256 byte stack and the compiler offers no software stack implementation.
Some of these variables are used as loop invariant or in place of local variables. Care has been taken
that no inner function accidentally overwrites global variables used by the calling functions.
Global variables in question are: `i`, `j`, `j2`, `k`, `m`, `dx`, `dy`, `ds`, `x`, `y`, `x1`, `y1`,
`x2`, `y2`, `field`, `attrField` and `b`.

## Screen Handling

Each screen has its one set of associated background/foreground tiles, tile maps and palettes. Some
are shared between these. A handler is defined for each screen. It handles the user interaction and
graphic output. All screen handlers are stored in `screenHandler` and being access in the main loop
of `main()` via the global variable `screen`. The enumeration values `S_TITLE`, `S_OPTIONS`, `S_GAME`,
`S_PAUSE` and `S_WINNER` correspond to the handlers of the same index in `screenHandler` for fast
selection. The values of the global variables `pad0` and `pad1` are updated before each call of the
screen handlers followed by a call to `WaitForVBlank()` to process the next frame.

# Debug

`debug.h` contains functions to create software breakpoints with assertions. These are used in various
places in the game and can be disabled via makefile parameter `NDEBUG=1`.

The software breakpoints are triggered via opcode `BRK`. Previously, the global pointer `debugMessage`
is set to point to the assertion string.

Use an emulator like [bsnes](https://github.com/bsnes-emu/bsnes) to debug with software breakpoints and
check the memory for the set address to see the failing assertion message.

# Known Issues

Sometimes the game freezes if it was built with `HAS_SFX=1` due to a race condition and/or memory bug in
the audio driver of pvsneslib. See [issue #226](https://github.com/alekmaul/pvsneslib/issues/226).

# License

All graphics and sounds are licensed under [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/).  
Code and other resources are being licensed under [3-clause BSD license](LICENSE).  
The used pvsneslib is being licensed under [MIT license](https://github.com/alekmaul/pvsneslib/blob/master/pvsneslib/pvsneslib_license.txt).

# Special Thanks

- fletchi for the character design
- [alekmaul](https://github.com/alekmaul) and the team behind [pvsneslib](https://github.com/alekmaul/pvsneslib)
- [Ulrich Hecht](http://code.google.com/p/snes-sdk/) for the initial SNES C SDK

# Disclaimer

This project is not affiliated with Nintendo K.K., or any of its affiliates. 


doc/Changelog.txt
version numbering schema:
M.N.P
| | +-- patch: increased if only bug fixed were incorporated
| +---- minor: increased if visual/audible changes were applied
+------ major: increased if elementary changes (from user's point of view) were made

1.1.0 (2023-07-29)
 - changed debugBreak and DEBUG_MSG to set the global variable debugMessage instead of the registers X and A
 - changed to pvsneslib and tcc develop branch to properly support const variable to ROM
 - changed bin/BombNBrake.sym format for direct use in BSNES via scripts/to-bsnes-sym.py
 - changed player walking animation sequence and frames slightly
 - fixed wrong explosion center tile for extending explosion intersection (i.e. same direction)
 - fixed updateOptionsScreen() writing to VRAM outside of VBlank in case of NDEBUG=0
 - fixed duplicate section name issue in main.o

1.0.0 (2023-07-17)
 - first release


src/debug.h
/**
* @file debug.h
* @author Daniel Starke
* @copyright Copyright 2023 Daniel Starke
* @date 2023-07-16
* @version 2023-07-21
*/
#ifndef _DEBUG_H_
#define _DEBUG_H_

#include <stddef.h>


/** Found in `debug.asm`. */
extern const char * debugMessage;


/** Concatenates two tokens. */
#define PP_CAT(x, y) PP_CAT_HELPER1(x, y)
#define PP_CAT_HELPER1(x, y) PP_CAT_HELPER2(x, y)
#define PP_CAT_HELPER2(x, y) x ## y


/** Returns a char[] literal of the given argument. */
#define PP_STR(x) PP_STR_HELPER(x)
#define PP_STR_HELPER(x) #x


/**
 * Calls the debugger with the given message.
 * 
 * @param[in] x - debug message to set
 */
#define DEBUG_MSG(x) \
	{debugMessage = PP_STR(__FILE__) ":" PP_STR(__LINE__) ":" x; debugBreak(); }


/**
 * @def ASSERT(x)
 * 
 * @param x - condition to assert
 */
/**
 * @def ASSERT_ARY_IDX(ary, idx)
 * 
 * @param ary - array to check (size needs to be known at compile time)
 * @param idx - array index to test
 */
#ifndef NDEBUG
#define ASSERT(x) \
	if ( ! (x) ) { DEBUG_MSG(PP_STR(x)); }
#define ASSERT_ARY_IDX(ary, idx) \
	if (idx >= (sizeof(ary) / sizeof(*(ary)))) { DEBUG_MSG("out of bounds"); }
#define ASSERT_ARY_PTR(ary, ptr) \
	ASSERT_ARY_IDX((ary), ((ptr) - (ary)))
#else /* NDEBUG */
#define ASSERT(x)
#define ASSERT_ARY_IDX(ary, idx)
#define ASSERT_ARY_PTR(ary, ptr)
#endif /* NDEBUG */


/**
 * Calls the break command.
 */
void debugBreak();


#endif /* _DEBUG_H_ */


src/hdr.asm
; @file hdr.asm
; @author Daniel Starke
; @copyright Copyright 2023 Daniel Starke
; @date 2023-07-03
; @version 2023-07-16


;==LoRom==

.MEMORYMAP                      ; Begin describing the system architecture.
  SLOTSIZE $8000                ; The slot is $8000 bytes in size. More details on slots later.
  DEFAULTSLOT 0                 ; There's only 1 slot in SNES, there are more in other consoles.
  SLOT 0 $8000                  ; Defines Slot 0's starting address.
  SLOT 1 $0 $2000
  SLOT 2 $2000 $E000
  SLOT 3 $0 $10000
.ENDME                          ; End MemoryMap definition

.ROMBANKSIZE $8000              ; Every ROM bank is 32 KBytes in size
.ROMBANKS 8                     ; 2 Mbits - Tell WLA we want to use 8 ROM Banks

.SNESHEADER
  ID "SNES"                     ; 1-4 letter string, just leave it as "SNES"

  NAME "Bomb'n'Break         "  ; Program Title - can't be over 21 bytes,
  ;    "123456789012345678901"  ; use spaces for unused bytes of the name.

  SLOWROM
  LOROM

  CARTRIDGETYPE $00             ; $00=ROM, $01=ROM+RAM, $02=ROM+SRAM, $03=ROM+DSP1, $04=ROM+RAM+DSP1, $05=ROM+SRAM+DSP1, $13=ROM+Super FX
  ROMSIZE $08                   ; $08=2 Megabits, $09=4 Megabits,$0A=8 Megabits,$0B=16 Megabits,$0C=32 Megabits
  SRAMSIZE $00                  ; $00=0 kilobits, $01=16 kilobits, $02=32 kilobits, $03=64 kilobits
.IFDEF USE_NTSC
  COUNTRY $01                   ; $01= U.S., $00=Japan, $02=Europe, $03=Sweden/Scandinavia, $04=Finland, $05=Denmark, $06=France, $07=Netherlands, $08=Spain, $09=Germany, $0A=Italy, $0B=China, $0C=Indonesia, $0D=Korea
.ELSE
  COUNTRY $02                   ; $01= U.S., $00=Japan, $02=Europe, $03=Sweden/Scandinavia, $04=Finland, $05=Denmark, $06=France, $07=Netherlands, $08=Spain, $09=Germany, $0A=Italy, $0B=China, $0C=Indonesia, $0D=Korea
.ENDIF
  LICENSEECODE $00              ; Just use $00
  VERSION $10                   ; $00 = 1.00, $01 = 1.01, etc.
.ENDSNES

.SNESNATIVEVECTOR               ; Define Native Mode interrupt vector table
  COP EmptyHandler
  BRK EmptyHandler
  ABORT EmptyHandler
  NMI VBlank
  IRQ EmptyHandler
.ENDNATIVEVECTOR

.SNESEMUVECTOR                  ; Define Emulation Mode interrupt vector table
  COP EmptyHandler
  ABORT EmptyHandler
  NMI EmptyHandler
  RESET tcc__start              ; where execution starts
  IRQBRK EmptyHandler
.ENDEMUVECTOR


src/utility.h
/**
* @file utility.h
* @author Daniel Starke
* @copyright Copyright 2023 Daniel Starke
* @date 2023-07-09
* @version 2023-07-15
*/
#ifndef _UTILITY_H_
#define _UTILITY_H_

#include <stdint.h>


/** Random seed for `lrng()`. Shall not be zero! */
extern uint32_t lrngSeed;


/**
 * Fill VRAM with the given word.
 *
 * @param[in] value - value to fill with
 * @param[in] address - VRAM address (word counting)
 * @param[in] size - VRAM size in number of bytes
 */
void dmaFillVramWord(const uint16_t value, const uint16_t address, const uint16_t size);


/**
 * Copy source bytes to VRAM low bytes.
 *
 * @param[in] source - source data address
 * @param[in] address - VRAM address (word counting)
 * @param[in] size - VRAM size in number of bytes
 */
void dmaCopyVramLowBytes(const uint8_t * source, const uint16_t address, const uint16_t size);


/**
 * Copy source bytes to VRAM high bytes.
 *
 * @param[in] source - source data address
 * @param[in] address - VRAM address (word counting)
 * @param[in] size - VRAM size in number of bytes
 */
void dmaCopyVramHighBytes(const uint8_t * source, const uint16_t address, const uint16_t size);


/**
 * Linear random number generator. A full period is (2^32)-1.
 *
 * @return random number in the range 0..65535 (16-bit)
 * @see https://www.jstatsoft.org/article/view/v008i14
 * @remarks Operates on the variable `rndSeed` to generate the next value. `lrngSeed` shall not be zero!
 * @remarks Takes 824 clock cycles (~0.6 scanlines).
 */
uint16_t lrng(void);


#endif /* _UTILITY_H_ */


src/utility.asm
; @file utility.asm
; @author Daniel Starke
; @copyright Copyright 2023 Daniel Starke
; @date 2023-07-07
; @version 2023-07-15

.equ REG_VMAIN    $2115
.equ REG_VMADDL   $2116
.equ REG_VMDATAL  $2118
.equ REG_VMDATAH  $2119
.equ REG_MDMAEN   $420B
.equ REG_DMAP0    $4300
.equ REG_BBAD0    $4301
.equ REG_A1T0L    $4302
.equ REG_A1B0     $4304
.equ REG_DAS0L    $4305

.RAMSECTION ".reg_utility7e" BANK $7E
; extern uint32_t lrngSeed;
lrngSeed:         DSW 2 ; random seed
.ends

.include "hdr.asm"
.accu 16
.index 16
.16bit


.section ".dmaFillVramWord_text" superfree
; void dmaFillVramWord(const uint16_t value, const uint16_t address, const uint16_t size);
dmaFillVramWord:
	php                ; push processor flags to stack (1 byte)
	                   ; stack:
	                   ; 9 | 2 byte size
	                   ; 7 | 2 byte address
	                   ; 5 | 2 byte value
	                   ; 1 | 4 byte return address
	                   ; 0 | 1 byte processor flags

	rep #$21           ; 16-bit accumulator, clear carry flag
	tsc                ; copy stack register to accumulator
	adc #5             ; address of the value on stack
	tax                ; copy accumulator to x register
	sta.l REG_A1T0L    ; source address
	lda 7,s
	sta.l REG_VMADDL   ; VRAM destination address (word addressed)
	lda 9,s
	sta.l REG_DAS0L    ; number of bytes to be written
	lda #$1808
	sta.l REG_DMAP0    ; fixed source address, byte increment source, operate on REG_VMDATAL

	sep #$20           ; 8-bit accumulator
	lda #$00
	sta.l REG_A1B0     ; source address in memory bank 0
    sta.l REG_VMAIN    ; increment VRAM address every low byte
	lda #1             ; turn on bit 1 (channel 0) of DMA
	sta.l REG_MDMAEN

	rep #$20           ; 16-bit accumulator
	txa                ; copy x register to accumulator
	inc a              ; next source byte
	sta.l REG_A1T0L    ; source address
	lda 7,s
	sta.l REG_VMADDL   ; VRAM destination address (word addressed)
	lda 9,s
	sta.l REG_DAS0L    ; number of bytes to be written

	sep #$20           ; 8-bit accumulator
	lda #$19
	sta.l REG_BBAD0    ; operate on REG_VMDATAH
	lda #$80
    sta.l REG_VMAIN    ; increment VRAM address every high byte
	lda #1             ; turn on bit 1 (channel 0) of DMA
	sta.l REG_MDMAEN

	plp                ; pull processor flags from stack (1 byte)
	rtl                ; return from subroutine long

.ends


.section ".dmaCopyVramLowBytes_text" superfree
; void dmaCopyVramLowBytes(const uint8_t * source, const uint16_t address, const uint16_t size);
dmaCopyVramLowBytes:
	php                ; push processor flags to stack (1 byte)
	                   ; stack:
	                   ; 11 | 2 byte size
	                   ;  9 | 2 byte address
	                   ;  5 | 4 byte source
	                   ;  1 | 4 byte return address
	                   ;  0 | 1 byte processor flags

	rep #$20           ; 16-bit accumulator
	lda 5,s
	sta.l REG_A1T0L    ; source address
	lda 9,s
	sta.l REG_VMADDL   ; VRAM destination address (word addressed)
	lda 11,s
	sta.l REG_DAS0L    ; number of bytes to be written
	lda #$1800
	sta.l REG_DMAP0    ; byte increment source, operate on REG_VMDATAL

	sep #$20           ; 8-bit accumulator
    lda 7,s
    sta.l REG_A1B0     ; bank address of the source
	lda #$00
    sta.l REG_VMAIN    ; increment VRAM address every high byte
	lda #1             ; turn on bit 1 (channel 0) of DMA
	sta.l REG_MDMAEN

	plp                ; pull processor flags from stack (1 byte)
	rtl                ; return from subroutine long

.ends


.section ".dmaCopyVramHighBytes_text" superfree
; void dmaCopyVramHighBytes(const uint8_t * source, const uint16_t address, const uint16_t size);
dmaCopyVramHighBytes:
	php                ; push processor flags to stack (1 byte)
	                   ; stack:
	                   ; 11 | 2 byte size
	                   ;  9 | 2 byte address
	                   ;  5 | 4 byte source
	                   ;  1 | 4 byte return address
	                   ;  0 | 1 byte processor flags

	rep #$20           ; 16-bit accumulator
	lda 5,s
	sta.l REG_A1T0L    ; source address
	lda 9,s
	sta.l REG_VMADDL   ; VRAM destination address (word addressed)
	lda 11,s
	sta.l REG_DAS0L    ; number of bytes to be written
	lda #$1900
	sta.l REG_DMAP0    ; byte increment source, operate on REG_VMDATAL

	sep #$20           ; 8-bit accumulator
    lda 7,s
    sta.l REG_A1B0     ; bank address of the source
	lda #$80
    sta.l REG_VMAIN    ; increment VRAM address every high byte
	lda #1             ; turn on bit 1 (channel 0) of DMA
	sta.l REG_MDMAEN

	plp                ; pull processor flags from stack (1 byte)
	rtl                ; return from subroutine long

.ends


.section ".lrng_text" superfree
; uint16_t lrng(void) {
; 	lrngSeed ^= lrngSeed >> 17;
; 	lrngSeed ^= lrngSeed << 15;
; 	lrngSeed ^= lrngSeed >> 23;
; 	return (uint16_t)lrngSeed;
; }
lrng:
	php                ; push processor flags to stack (1 byte)
	rep #$20           ; 16-bit accumulator
	; lrngSeed ^= lrngSeed >> 17;
	lda.w lrngSeed + 2 ; high word to accumulator
	lsr A              ; shift accumulator right; set carry flag with shifted bit
	eor.w lrngSeed     ; xor accumulator with low word
	sta.w lrngSeed     ; store accumulator in low word
	; lrngSeed ^= lrngSeed << 15;
	lda.w lrngSeed     ; low word to accumulator
	ror A              ; rotate accumulator right; use previously set carry flag for the MSB; set carry flag with removed bit
	eor.w lrngSeed + 2 ; xor accumulator with high word
	sta.w lrngSeed + 2 ; store accumulator in high word
	lda.w #0           ; 0 to accumulator
	ror A              ; set MSB from carry flag
	eor.w lrngSeed     ; xor accumulator with low word
	sta.w lrngSeed     ; store accumulator in low word
	; lrngSeed ^= lrngSeed >> 23;
	sep #$20           ; 8-bit accumulator
	lda.w lrngSeed + 2 ; load low byte in high word to accumulator
	asl A              ; shift accumulator left; set carry flag with shifted bit
	lda.w lrngSeed + 3 ; load high byte in high word to accumulator
	rol A              ; rotate accumulator left; use previously set carry flag for the LSB; set carry flag with removed bit
	eor.w lrngSeed     ; xor accumulator with low byte in low word
	sta.w lrngSeed     ; store accumulator in low byte of low word
	lda #0             ; 0 to accumulator
	rol A              ; set LSB from carry flag
	eor.w lrngSeed + 1 ; xor accumulator with high byte in low word
	sta.w lrngSeed + 1 ; store accumulator in high byte of low word
	; return (uint16_t)lrngSeed;
	rep #$20           ; 16-bit accumulator
	lda.w lrngSeed     ; load low word to accumulator
	sta.b tcc__r0      ; store accumulator in return register
	plp                ; pull processor flags from stack (1 byte)
	rtl                ; return from subroutine long

.ends


src/data.asm
; @file data.asm
; @author Daniel Starke
; @copyright Copyright 2023 Daniel Starke
; @date 2023-07-03
; @version 2023-07-13

.include "hdr.asm"


.section ".rodata1" superfree

bg1Tiles:
.incbin "bg1.chr"
bg1TilesEnd:

fg1Tiles:
.incbin "fg1.chr"
fg1TilesEnd:

fg2Tiles:
.incbin "fg2.chr"
fg2TilesEnd

p12Tiles:
.incbin "p12.chr"
p12TilesEnd

.ends


.section ".rodata2" superfree

bg1Map:
.incbin "bg1.map"
bg1MapEnd:

bg2Map:
.incbin "bg2.map"
bg2MapEnd:

fg1Map:
.incbin "fg1.map"
fg1MapEnd:

optionsMap:
.incbin "options.map"
optionsMapEnd:

fieldMap:
.incbin "field.map"
fieldMapEnd:

bg1Pal:
.incbin "bg1.pal"
bg1PalEnd:

fg1Pal:
.incbin "fg1.pal"
fg1PalEnd:

fg2Pal:
.incbin "fg2.pal"
fg2PalEnd:

p12Pal:
.incbin "p12.pal"
p12PalEnd:

.ends


.ifdef HAS_SFX
.section ".rodata3" superfree

sfx1:
.incbin "sfx1.brr"
sfx1End:

.ends
.endif